### ISSUE TITLE
Take latest

### ISSUE THREAD

**Issue by kishorevarma** [jeffbski/redux-logic#5]
Hi ,

I am checking the simple example [http://jsfiddle.net/jeffbski/954g5n7h](example) , and its using `latest: true`, that means it has to cancel previous requests as per documentation.

But I observed its not cancelling the previous requests, instead its waiting until all the previous requests are completed, is it correct way , am I understanding the feature wrongly ?

**Comment by jeffbski** [comment]
Thanks for asking. I'll have to find the words to make this clearer in the docs.

In order to have redux-logic cancel or take latest actively abort in-flight xhr requests you can use the rxjs ajax (instead of axios) and dispatch the observable. Then when cancel or take latest fires, those outstanding xhr requests are truly cancelled.

Here is a jsfiddle demonstrating how this works, you can see the xhr cancellations in the chrome network console.

http://jsfiddle.net/jeffbski/0fu407na/

or alternatively using the new processOptions feature with dispatchReturn, successType, and failType then the code is simplified further:

http://jsfiddle.net/jeffbski/d4p6oo0d/


If you are not using the rxjs ajax but something like axios then there is not a way to truly abort in-flight requests, all we can do is simply ignore them when they come back.

In general, unless you give redux-logic an observable, then it can't automatically abort  from a cancel/latest. Normally this is ok since things will end when the responses come back, and in many cases just not dispatching the results is sufficient. However as you noticed, it could mean that additional requests are delayed since the browser often limits the number of requests per host. redux-logic itself doesn't impose any delays waiting for previous requests, but the browser logic could and that's what you were observing.

redux-logic also provides a way to detect a cancellation/latest in which you could manually abort or cleanup. One of the other properties available in the first arg provided to the hooks is cancelled$. It is an observable that will emit next if a cancel/latest occurred. It also will complete when the logic is done (cancelled or normal completion). So you can subscribe to that to abort a request or do additional cleanup. Unfortunately axios/fetch don't support cancellation yet, so this wouldn't really help for simple requests. It is useful for things like web sockets.

So if you do want true xhr abort cancellation and not just non-dispatching of the results then the best thing is to use rxjs ajax and dispatch the observable so cancels will go all the way to the source and abort the xhr requests. 

I think I will change the cancel examples to use rxjs ajax and explain that if you use other mechanisms like axios that the xhr requests are not aborted but only the results ignored on return.

Hopefully this helps.

**Comment by shauntrennery** [comment]
Which is the recommended xhr lib to use? rxjs ajax, axios or fetch? True request cancellation is not a requirement of my implementation.

**Comment by jeffbski** [comment]
@shauntrennery it seems to me that axios and fetch are both quite popular and I've never run into any issues with either.

I think it comes down to personal choice, but here's my recommendation:

If you don't need xhr abort/cancel then axios or fetch. axios automatically parses json so that eliminates an extra `then` statement in your code, so I have been using that. Fetch would be my second choice. 

Also axios is isomporphic or universal in that it will run on browser or server with node.js. There is isomorphic-fetch that will do the same with fetch.

And if I need true xhr abort cancellation then RxJS ajax.

**Comment by nicholaswyoung** [comment]
Personally, I use [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API), [polyfilled in the browser](https://github.com/github/fetch) where necessary. `isomorphic-fetch` is awesome if you must have universal rendering.

Then, I wrap the standards-compliant Promise API: [`Observable.fromPromise`](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/frompromise.md). That way, I'm leaning on native browser APIs where possible, at least until Observables are part of the spec.

**Comment by jeffbski** [comment]
@nicholaswyoung is there any objective reasons that you like fetch better than axios?

I guess the fact that it is being implemented by the browser vendors means that eventually we wouldn't need the polyfill and it should support the widest set of use cases. 

I know fetch is popular, so I am trying to learn exactly why it is so popular. 

Not trolling here, just wanting to learn and know the trade offs between things.

**Comment by nicholaswyoung** [comment]
@jeffbski Here goes. I'll try to be brief. ;)

**tldr;** `fetch` is popular because it provides the flexibility of a relatively low-level API without the cruft and configuration of `XMLHTTPRequest`. Whether it moves to support the whatwg streams proposal or not, it's still the result of [significant effort in the standards community](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) and can't be easily ignored.

Axios is great, but I believe it's time has come and gone. It sits on top of XHR, the implementations of which (last I checked, anyway) varied between Node and the browser. 

Granted, I've only worked with Axios a handful of times, but XHR will likely be retired at some point. The expectation is that `fetch` will replace it. If I can choose between a library or eventual native implementation, I'll bet on the latter most any day.

Since `fetch` is a proposed native API (which has [decent support already]http://caniuse.com/#feat=fetch)), and will eventually provide the freedom to eventually trim my compiled package size (or even conditionally include the polyfill), `fetch` is a more viable approach for the future, IMO.

Plus, wrapping a Promise in an Observable is a fairly lightweight process. In fact, it's a decent pattern, since Promises represent a single async action, whereas Observables represent a ongoing stream.

**Comment by jeffbski** [comment]
@nicholaswyoung Thanks. Very clear explanation, that helps a bunch.

**Comment by kishorevarma** [comment]
Thanks @jeffbski , following is the sentence in READ.me  that I mentioned in my previous message.
" It also uses take latest feature that if additional FETCH_POLLS actions come in before this completes, it will simply cancel the previous requests and just use the latest."

**Comment by jeffbski** [comment]
@kishorevarma Thanks, I've updated that statement in the README to now say "It also uses `take latest` feature that if additional FETCH_POLLS actions come in before this completes, it will ignore the outdated requests."

I've also added more details in the advanced section of the API doc to help explain the situation. 

https://github.com/jeffbski/redux-logic/blob/master/docs/api.md#cancellation--take-latest---xhr-aborting

I was reading about http/2 today and one of the benefits of moving to that is that it lifts the limits of requests since they are all multiplexed in one connection, so as more servers move to http2 then this will become less of an issue. I mean it't nice that we can get full cancellation using RxJS ajax which saves bandwidth too, but as for latency it will probably become less of an issue as things move to http-2.

I've added a couple full examples in addition to the jsfiddle ones showing RxJS ajax use, hopefully that will help people that need it. 

Let me know if there is anything else you would suggest.

**Comment by jeffbski** [comment]
I guess I'll close this. If you think of any additional clarifications we can make let me know. Thanks.

**Comment by kishorevarma** [comment]
Thanks @jeffbski
